File: server/utils/winget/tree.ts
import { env } from "std-env";

import { cacheStorage } from "../storage";
import {
  WINGET_CACHE_PREFIX,
  WINGET_GITHUB_API_BASE,
  WINGET_GITHUB_BRANCH,
  WINGET_GITHUB_REPO,
  WINGET_MANIFESTS_SHA_KEY,
  WINGET_UPDATE_INTERVAL,
} from "./constants";
import type { WinGetGitHubTreeResponse } from "./types";

/**
 * Fetch GitHub tree data by SHA or branch
 */
export async function getGitHubTree(
  treeSha: string = WINGET_GITHUB_BRANCH,
  recursive: boolean = false,
): Promise<WinGetGitHubTreeResponse> {
  const url = `${WINGET_GITHUB_API_BASE}/repos/${WINGET_GITHUB_REPO}/git/trees/${treeSha}${recursive ? "?recursive=1" : ""}`;

  const headers: HeadersInit = { "User-Agent": "Funish Nexus" };
  const token = env.GITHUB_TOKEN;
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  const response = await fetch(url, { headers });

  if (!response.ok) {
    throw new Error(`Failed to fetch GitHub tree: ${response.statusText}`);
  }

  return (await response.json()) as WinGetGitHubTreeResponse;
}

/**
 * Fetch and cache GitHub tree paths (optimized storage)
 * @param treeSha - The SHA of the tree
 * @param cacheSuffix - Cache suffix for this tree (e.g. "manifests/m")
 * @returns Array of file paths
 */
export async function getGitHubTreePaths(treeSha: string, cacheSuffix: string): Promise<string[]> {
  const normalizedSuffix = cacheSuffix.replace(/\//g, "-");
  const cacheKey = `${WINGET_CACHE_PREFIX}/${normalizedSuffix}`;

  // Check cache metadata
  const meta = await cacheStorage.getMeta(cacheKey);
  const now = new Date();

  if (meta?.mtime) {
    const cacheAge = (now.getTime() - new Date(meta.mtime).getTime()) / 1000;
    if (cacheAge < WINGET_UPDATE_INTERVAL) {
      const cached = await cacheStorage.getItem(cacheKey);
      if (cached && Array.isArray(cached)) {
        return cached as string[];
      }
    }
  }

  // Fetch tree data
  const treeData = await getGitHubTree(treeSha, true);

  // Extract only paths from tree items
  const paths = treeData.tree.map((item) => item.path);

  // Cache the paths array
  await cacheStorage.setItem(cacheKey, paths);
  await cacheStorage.setMeta(cacheKey, { mtime: new Date() });

  return paths;
}

/**
 * Get all letter directory SHAs from manifests (a-z, 0-9)
 */
export async function getLetterDirectoryShas(): Promise<Map<string, string>> {
  const cacheKey = `${WINGET_CACHE_PREFIX}/letter-shas.json`;

  // Check cache
  const meta = await cacheStorage.getMeta(cacheKey);
  const now = new Date();

  if (meta?.mtime) {
    const cacheAge = (now.getTime() - new Date(meta.mtime).getTime()) / 1000;
    if (cacheAge < WINGET_UPDATE_INTERVAL) {
      const cached = await cacheStorage.getItem(cacheKey);
      if (cached) {
        return new Map(Object.entries(cached as Record<string, string>));
      }
    }
  }

  // Get manifests SHA (cached 10-min TTL)
  const shaMeta = await cacheStorage.getMeta(WINGET_MANIFESTS_SHA_KEY);
  let manifestsSha: string;

  if (shaMeta?.mtime) {
    const shaAge = (now.getTime() - new Date(shaMeta.mtime).getTime()) / 1000;
    if (shaAge < WINGET_UPDATE_INTERVAL) {
      const cached = await cacheStorage.getItem(WINGET_MANIFESTS_SHA_KEY);
      if (cached && typeof cached === "string") {
        manifestsSha = cached;
      } else {
        manifestsSha = await fetchManifestsSha();
      }
    } else {
      manifestsSha = await fetchManifestsSha();
    }
  } else {
    manifestsSha = await fetchManifestsSha();
  }

  const manifestsTree = await getGitHubTree(manifestsSha, false);

  const letterShas = new Map<string, string>();

  for (const item of manifestsTree.tree) {
    if (item.type === "tree" && item.path.length === 1 && /[a-z0-9]/.test(item.path)) {
      letterShas.set(item.path, item.sha);
    }
  }

  if (letterShas.size === 0) {
    throw new Error("No letter directories found in manifests");
  }

  // Cache as plain object (Map is not serializable)
  await cacheStorage.setItem(cacheKey, Object.fromEntries(letterShas));
  await cacheStorage.setMeta(cacheKey, { mtime: new Date() });

  return letterShas;
}

/**
 * Fetch the manifests directory SHA from the root tree.
 */
export async function fetchManifestsSha(): Promise<string> {
  const rootTree = await getGitHubTree(WINGET_GITHUB_BRANCH, false);
  const manifestsItem = rootTree.tree.find(
    (item) => item.path === "manifests" && item.type === "tree",
  );

  if (!manifestsItem) {
    throw new Error("manifests directory not found in repository");
  }

  await cacheStorage.setItem(WINGET_MANIFESTS_SHA_KEY, manifestsItem.sha);
  await cacheStorage.setMeta(WINGET_MANIFESTS_SHA_KEY, { mtime: new Date() });

  return manifestsItem.sha;
}

File: src/winget/utils/tree.rs
//! GitHub Trees API access with TTL caching (mirrors winget/tree.ts).
//!
//! Discovers manifest file paths under `manifests/<letter>/...` and caches the
//! letter-directory SHAs and tree paths with a 10-minute TTL.

use std::collections::HashMap;
use std::time::Duration;

use anyhow::Result;
use serde::Deserialize;

use crate::storage::SharedStorage;

use super::constants::*;

#[derive(Debug, Deserialize)]
struct TreeItem {
    path: String,
    #[allow(dead_code)]
    mode: String,
    #[serde(rename = "type")]
    item_type: String,
    sha: String,
    #[allow(dead_code)]
    url: String,
}

#[derive(Debug, Deserialize)]
struct TreeResponse {
    #[allow(dead_code)]
    sha: String,
    tree: Vec<TreeItem>,
}

/// Fetch a GitHub tree by SHA or branch (mirrors getGitHubTree).
async fn get_github_tree(tree_sha: &str, recursive: bool) -> Result<TreeResponse> {
    let url = format!(
        "{WINGET_GITHUB_API_BASE}/repos/{WINGET_GITHUB_REPO}/git/trees/{tree_sha}{}",
        if recursive { "?recursive=1" } else { "" }
    );
    let mut req = super::http::HTTP_CLIENT
        .get(&url)
        .timeout(Duration::from_secs(30));
    if let Ok(token) = std::env::var("GITHUB_TOKEN") {
        req = req.header("Authorization", format!("Bearer {token}"));
    }
    let resp = req.send().await?;
    if !resp.status().is_success() {
        anyhow::bail!("Failed to fetch GitHub tree: {}", resp.status());
    }
    Ok(resp.json().await?)
}

/// Cached, recursive tree file paths (mirrors getGitHubTreePaths).
pub async fn get_github_tree_paths(
    storage: &SharedStorage,
    tree_sha: &str,
    cache_suffix: &str,
) -> Result<Vec<String>> {
    let normalized = cache_suffix.replace('/', "-");
    let cache_key = format!("{WINGET_CACHE_PREFIX}/{normalized}");

    if cache_fresh(storage, &cache_key).await
        && let Some(data) = storage.get_raw(&cache_key).await
        && let Ok(paths) = serde_json::from_slice::<Vec<String>>(&data)
    {
        return Ok(paths);
    }

    let tree = get_github_tree(tree_sha, true).await?;
    let paths: Vec<String> = tree.tree.into_iter().map(|i| i.path).collect();

    storage
        .set_raw(&cache_key, &serde_json::to_vec(&paths)?)
        .await;
    set_mtime(storage, &cache_key).await;
    Ok(paths)
}

/// Cached letter-directory SHAs (a-z, 0-9) under manifests/ (mirrors getLetterDirectoryShas).
pub async fn get_letter_directory_shas(storage: &SharedStorage) -> Result<HashMap<String, String>> {
    let cache_key = format!("{WINGET_CACHE_PREFIX}/letter-shas.json");

    if cache_fresh(storage, &cache_key).await
        && let Some(data) = storage.get_raw(&cache_key).await
        && let Ok(map) = serde_json::from_slice::<HashMap<String, String>>(&data)
    {
        return Ok(map);
    }

    let manifests_sha = fetch_manifests_sha(storage).await?;
    let tree = get_github_tree(&manifests_sha, false).await?;

    let mut shas = HashMap::new();
    for item in &tree.tree {
        if item.item_type == "tree"
            && item.path.len() == 1
            && item.path.chars().all(|c| c.is_ascii_alphanumeric())
        {
            shas.insert(item.path.clone(), item.sha.clone());
        }
    }

    if shas.is_empty() {
        anyhow::bail!("No letter directories found in manifests");
    }

    storage
        .set_raw(&cache_key, &serde_json::to_vec(&shas)?)
        .await;
    set_mtime(storage, &cache_key).await;
    Ok(shas)
}

/// Cached SHA of the manifests/ directory (mirrors fetchManifestsSha).
pub async fn fetch_manifests_sha(storage: &SharedStorage) -> Result<String> {
    if cache_fresh(storage, WINGET_MANIFESTS_SHA_KEY).await
        && let Some(data) = storage.get_raw(WINGET_MANIFESTS_SHA_KEY).await
        && let Ok(s) = String::from_utf8(data)
    {
        return Ok(s);
    }

    let root = get_github_tree(WINGET_GITHUB_BRANCH, false).await?;
    let manifests = root
        .tree
        .iter()
        .find(|i| i.path == "manifests" && i.item_type == "tree")
        .ok_or_else(|| anyhow::anyhow!("manifests directory not found in repository"))?;

    let sha = manifests.sha.clone();
    storage
        .set_raw(WINGET_MANIFESTS_SHA_KEY, sha.as_bytes())
        .await;
    set_mtime(storage, WINGET_MANIFESTS_SHA_KEY).await;
    Ok(sha)
}

/// Whether a cache entry is present and younger than the update interval.
async fn cache_fresh(storage: &SharedStorage, key: &str) -> bool {
    let Some(meta) = storage.get_meta(key).await else {
        return false;
    };
    let Some(mtime) = meta.mtime else {
        return false;
    };
    let Ok(ts) = chrono::DateTime::parse_from_rfc3339(&mtime) else {
        return false;
    };
    let age = chrono::Utc::now()
        .signed_duration_since(ts.with_timezone(&chrono::Utc))
        .num_seconds();
    age < WINGET_UPDATE_INTERVAL_SECS
}

/// Stamp a cache entry's mtime to "now".
async fn set_mtime(storage: &SharedStorage, key: &str) {
    storage
        .set_meta(
            key,
            &crate::storage::CacheMeta {
                mtime: Some(chrono::Utc::now().to_rfc3339()),
                ..Default::default()
            },
        )
        .await;
}
